### ISSUE TITLE
[Question] Using as an E2E suite with TypeScript

### ISSUE THREAD

**Issue by naknode** [microsoft/playwright#12896]
I have a form I want to E2E test and make sure it saves fine. I use Playwright to run the test. It's written all in TypeScript. The form has many dropdowns which are filled in via `enums` with their `mappers`.

The problem is that I use Dropdowns (and Vuetify so not a native dropdown).  What's the best way to go about this? Should I import the `enum` and make Playwright select that or is it better practice to write Playwright to go inside the dropdown and select a random item?

Basically... I should try to not use as much of the source code in the E2E as possible -- correct?

**Comment by pavelfeldman** [comment]
That's up to you, but it is typically good to abstract from the implementation and test against explicit expectations. Otherwise, 'Submit' button can be renamed to 'Destroy' and your tests will keep passing.

**Comment by naknode** [comment]
@pavelfeldman Yeah -- that makes sense. I suppose what I'm having trouble with is how I'm going to split it up to accommodate the E2E tests.

What I am doing now is that I need to create an object of the values I'm going to check against (with `expect`) versus what I'm going to use to fill the form with.

So, inside `test()`, I have something like:

`edit-user.spec.ts:`
```typescript
  test.only("can we update the 'General Info' pane of a user profile edit page", async ({ page }) => {
    const updateUserPage = new UpdateUserPage(page,  "uuid-of-user");
    await updateUserPage.start();

    // General Info
    await updateUserPage.selectFromDropdown("userLevel", updateUserPage.userProfile.userLevel);
    await updateUserPage.selectFromDropdown("Theme", updateUserPage.userProfile.theme);

    // Save button
    await updateUserPage.page.locator('text=Save').click();

    // Wait for green toasty at top to confirm saving
    await updateUserPage.page.locator('.successful-save').waitFor();
  });
```

`update-user.page.ts:`
```typescript
export class UpdateUserPage {
  readonly page: Page;
  readonly uuid: string;
  public userProfile?: any;

  /**
   * Constructs a new page object model for creating an Item Master
   * @param page Page
   */
  constructor(page: Page, uuid?: string) {
    this.page = page;
    this.uuid = uuid;
  }

  async start() {
    await this.page.goto("user/profile/" + this.uuid);
    this.userProfile = await this.generateGeneralInfoData();
  }

  async generateGeneralInfoData() {
    return {
      uuid: this.uuid,
      interests:  [...faker.random.words(3).split(" ")],
      title: faker.git.commitMessage(),
      userLevel: await this.selectRandomDropdown("User Level"),
      theme: await this.selectRandomDropdown("Theme"),
      description: faker.random.word(),
      contractId: faker.datatype.number({min: 10, max: 100}).toString(),
    }
  }

  fillOutTextField(label: string, value: string) {
    return this.page.locator(`input[aria-label="${label}"]`).fill(value);
  }
}
```

It seems like there would be an easier way to create the form of the page you're going to fill out and use that form to check with against the `expect()` with what you see on the screen.

It seems like I'm creating too much work.

**Comment by dgozman** [comment]
@naknode That looks like a fine way to structure your tests, especially if you like the POM pattern.

One minor note - I'd suggest to use [locator assertions](https://playwright.dev/docs/test-assertions):

```js
// replace this:
expect(await (updateUserPage.page.locator('[data-user-theme]')).textContent).toBe(updateUserPage.userProfile.theme);
// with this:
await expect(updateUserPage.page.locator('[data-user-theme]')).toHaveText(updateUserPage.userProfile.theme);
```

**Comment by naknode** [comment]
@dgozman Thank you.

Is there a reason why this doesn't work? 

`BasePOM.ts`:
```typescript
  async getInput(label: string): Promise<Locator> {
    return this.page.locator(`input[aria-label="${label}"]`);
  }
```
and `edit.spec.ts`:

```typescript
expect(updateUserPage.getInput("Theme")).toHaveValue(updateUserPage. userProfile.theme);
```'

`updateUserPage` is instantiated in `edit.spec.ts`: and `UpdateUserPage.page.ts` extends `BasePOM.ts` which has the `getInput` function. Perhaps this is bad practice?

**Comment by dgozman** [comment]
You need two changes: make `getInput` sync instead of async, and add `await` before expect:

```js
getInput(label: string): Locator {
  return this.page.locator(`input[aria-label="${label}"]`);
}

await expect(updateUserPage.getInput("Theme")).toHaveValue(updateUserPage. userProfile.theme);
```

**Comment by mxschmitt** [comment]
Closing since all questions are answered. For further questions please open new issues.
